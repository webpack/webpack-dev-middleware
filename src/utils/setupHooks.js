export default function setupHooks(context) {
  function invalid() {
    if (context.state) {
      context.logger.info('Compiling...');
    }

    // We are now in invalid state
    // eslint-disable-next-line no-param-reassign
    context.state = false;
    // eslint-disable-next-line no-param-reassign, no-undefined
    context.stats = undefined;
  }

  function done(stats) {
    // We are now on valid state
    // eslint-disable-next-line no-param-reassign
    context.state = true;
    // eslint-disable-next-line no-param-reassign
    context.stats = stats;

    // Do the stuff in nextTick, because bundle may be invalidated if a change happened while compiling
    process.nextTick(() => {
      const { state, callbacks } = context;

      // Check if still in valid state
      if (!state) {
        return;
      }

      // eslint-disable-next-line no-console
      console.log(stats.toString());

      // eslint-disable-next-line no-param-reassign
      context.callbacks = [];

      // Execute callback that are delayed
      callbacks.forEach((callback) => {
        callback(stats);
      });
    });
  }

  context.compiler.hooks.watchRun.tap('webpack-dev-middleware', invalid);
  context.compiler.hooks.invalid.tap('webpack-dev-middleware', invalid);
  (context.compiler.webpack
    ? context.compiler.hooks.afterDone
    : context.compiler.hooks.done
  ).tap('webpack-dev-middleware', done);
}
