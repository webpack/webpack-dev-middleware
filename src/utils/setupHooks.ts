import type webpack from 'webpack';
import type { MultiStats, WebpackDevMiddlewareContext } from '../types';
import { isMultiCompiler } from './isMultiCompiler';

export default function setupHooks(context: WebpackDevMiddlewareContext): void {
  function invalid() {
    if (context.state) {
      context.logger.info('Compiling...');
    }

    // We are now in invalid state
    context.state = false;
    context.stats = null;
  }

  function done(stats: webpack.Stats | MultiStats) {
    // We are now on valid state
    context.state = true;
    context.stats = stats;

    // Do the stuff in nextTick, because bundle may be invalidated if a change happened while compiling
    process.nextTick(() => {
      const { state, compiler, callbacks, logger } = context;

      // Check if still in valid state
      if (!state) {
        return;
      }

      // Print webpack output
      const printStats = (
        childCompiler: webpack.Compiler,
        childStats: webpack.Stats
      ) => {
        const statsString = childStats.toString(childCompiler.options.stats);
        const name = childCompiler.options.name
          ? `Child "${childCompiler.options.name}": `
          : '';

        if (statsString.length) {
          if (childStats.hasErrors()) {
            logger.error(`${name}${statsString}`);
          } else if (childStats.hasWarnings()) {
            logger.warn(`${name}${statsString}`);
          } else {
            logger.info(`${name}${statsString}`);
          }
        }

        let message = `${name}Compiled successfully.`;

        if (childStats.hasErrors()) {
          message = `${name}Failed to compile.`;
        } else if (childStats.hasWarnings()) {
          message = `${name}Compiled with warnings.`;
        }

        logger.info(message);
      };

      if (isMultiCompiler(compiler)) {
        compiler.compilers.forEach((compilerFromMultiCompileMode, index) => {
          printStats(
            compilerFromMultiCompileMode,
            (stats as MultiStats).stats[index]
          );
        });
      } else {
        printStats(compiler, stats as webpack.Stats);
      }

      context.callbacks = [];

      // Execute callback that are delayed
      callbacks.forEach(
        (
          callback:
            | ((stats: webpack.Stats) => void)
            | ((stats: MultiStats) => void)
        ) => {
          callback(stats as webpack.Stats & MultiStats);
        }
      );
    });
  }

  context.compiler.hooks.watchRun.tap('DevMiddleware', invalid);
  context.compiler.hooks.invalid.tap('DevMiddleware', invalid);
  context.compiler.hooks.done.tap('DevMiddleware', done);
}
