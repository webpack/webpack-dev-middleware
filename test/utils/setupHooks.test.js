import setupHooks from "../../src/utils/setupHooks";

// Suppress unnecessary stats output
global.console.log = jest.fn();

describe("setupHooks", () => {
  let context;
  const watchRunHook = jest.fn();
  const invalidHook = jest.fn();
  const doneHook = jest.fn();
  const loggerLog = jest.fn();
  const loggerInfo = jest.fn();
  const loggerWarn = jest.fn();
  const loggerError = jest.fn();
  let nextTick;

  const cb1 = jest.fn();
  const cb2 = jest.fn();

  beforeEach(() => {
    nextTick = jest.spyOn(process, "nextTick").mockImplementation(() => {});
    context = {
      options: {},
      compiler: {
        hooks: {
          watchRun: {
            tap: watchRunHook,
          },
          invalid: {
            tap: invalidHook,
          },
          done: {
            tap: doneHook,
          },
        },
        options: { stats: {} },
      },
      logger: {
        log: loggerLog,
        info: loggerInfo,
        warn: loggerWarn,
        error: loggerError,
      },
      callbacks: [cb1, cb2],
    };
  });

  afterEach(() => {
    watchRunHook.mockClear();
    invalidHook.mockClear();
    doneHook.mockClear();
    loggerInfo.mockClear();
    loggerWarn.mockClear();
    loggerError.mockClear();
    nextTick.mockClear();
    cb1.mockClear();
    cb2.mockClear();
  });

  it("taps watchRun, invalid, and done", () => {
    setupHooks(context);
    expect(watchRunHook.mock.calls.length).toEqual(1);
    expect(invalidHook.mock.calls.length).toEqual(1);
    expect(doneHook.mock.calls.length).toEqual(1);
  });

  it("watchRun hook invalidates", () => {
    setupHooks(context);
    // this calls invalidate
    watchRunHook.mock.calls[0][1]();
    expect(context.state).toEqual(false);
    expect(context.stats).toBeUndefined();
    expect(loggerInfo.mock.calls.length).toEqual(0);
  });

  it("invalid hook invalidates", () => {
    setupHooks(context);
    // this calls invalidate
    invalidHook.mock.calls[0][1]();
    expect(context.state).toEqual(false);
    expect(context.stats).toBeUndefined();
    expect(loggerInfo.mock.calls.length).toEqual(0);
  });

  it("logs if state is set on invalidate", () => {
    context.state = true;
    setupHooks(context);
    // this calls invalidate
    invalidHook.mock.calls[0][1]();
    expect(context.state).toEqual(false);
    expect(context.stats).toBeUndefined();
    expect(loggerLog.mock.calls[0][0]).toEqual("Compilation starting...");
  });

  it("sets state, then logs stats and handles callbacks on nextTick from done hook", () => {
    setupHooks(context);
    doneHook.mock.calls[0][1]({
      toString: jest.fn(() => "statsString"),
      hasErrors: jest.fn(() => false),
      hasWarnings: jest.fn(() => false),
    });
    expect(context.stats).toBeTruthy();
    expect(context.state).toBeTruthy();
    expect(nextTick.mock.calls.length).toEqual(1);

    nextTick.mock.calls[0][0]();
    expect(loggerInfo.mock.calls).toMatchSnapshot();
    expect(loggerError.mock.calls.length).toEqual(0);
    expect(loggerWarn.mock.calls.length).toEqual(0);

    expect(cb1.mock.calls[0][0]).toEqual(context.stats);
    expect(cb2.mock.calls[0][0]).toEqual(context.stats);
  });

  it("stops on done if invalidated before nextTick", () => {
    setupHooks(context);
    doneHook.mock.calls[0][1]("stats");
    expect(context.stats).toEqual("stats");
    expect(context.state).toBeTruthy();
    expect(nextTick.mock.calls.length).toEqual(1);
    context.state = false;
    nextTick.mock.calls[0][0]();
    expect(loggerInfo.mock.calls.length).toEqual(0);
  });

  it("handles multi compiler", () => {
    context.compiler.compilers = [
      {
        options: {
          name: "comp1",
          stats: {},
        },
      },
      {
        options: {
          name: "comp2",
          stats: {},
        },
      },
    ];
    setupHooks(context);
    doneHook.mock.calls[0][1]({
      stats: [
        {
          toString: jest.fn(() => "statsString1"),
          hasErrors: jest.fn(() => true),
          hasWarnings: jest.fn(() => false),
        },
        {
          toString: jest.fn(() => "statsString2"),
          hasErrors: jest.fn(() => false),
          hasWarnings: jest.fn(() => true),
        },
      ],
    });
    expect(context.stats).toBeTruthy();
    expect(context.state).toBeTruthy();
    expect(nextTick.mock.calls.length).toEqual(1);

    nextTick.mock.calls[0][0]();
    expect(loggerInfo.mock.calls).toMatchSnapshot();
    expect(loggerError.mock.calls).toMatchSnapshot();
    expect(loggerWarn.mock.calls).toMatchSnapshot();

    expect(cb1.mock.calls[0][0]).toEqual(context.stats);
    expect(cb2.mock.calls[0][0]).toEqual(context.stats);
  });
});
