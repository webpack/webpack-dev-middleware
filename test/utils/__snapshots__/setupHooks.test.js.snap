// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupHooks handles multi compiler 1`] = `Array []`;

exports[`setupHooks handles multi compiler 2`] = `Array []`;

exports[`setupHooks handles multi compiler 3`] = `Array []`;

exports[`setupHooks sets state, then logs stats and handles callbacks on nextTick from done hook 1`] = `Array []`;
